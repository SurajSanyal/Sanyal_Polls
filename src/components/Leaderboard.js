import { connect } from "react-redux";

const Leaderboard = ({ sortedUsers }) => {
  return (
    <table className="table-auto mx-auto">
      <thead>
        <tr>
          <th>User</th>
          <th>Questions Asked</th>
          <th>Questions Answered</th>
        </tr>
      </thead>

      <tbody>
        {sortedUsers.map((user) => {
          return (
            <tr key={user.id}>
              <td className="grid justify-center">
                <div className="flex items-center justify-self-auto">
                  <img src={user.avatarURL} alt="Profile" className="h-10" />
                  <span>{user.name}</span>
                </div>
              </td>
              <td>{user.questions.length}</td>
              <td>{Object.keys(user.answers).length}</td>
            </tr>
          );
        })}
      </tbody>
    </table>
  );
};

// Access store to sort users from most to least score
const mapStateToProps = ({ users }) => {
  // sorting users by #q's + #a's.
  // returns array of usernames
  const sortedUsernames = Object.keys(users).sort((a, b) => {
    const scoreA =
      Object.keys(users[a].answers).length + users[a].questions.length;
    const scoreB =
      Object.keys(users[b].answers).length + users[b].questions.length;

    return scoreB - scoreA;
  });

  return {
    sortedUsers: sortedUsernames.map((username) => users[username]),
  };
};

export default connect(mapStateToProps)(Leaderboard);
