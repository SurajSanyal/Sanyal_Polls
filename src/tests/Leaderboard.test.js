import React from "react";
import { Provider } from "react-redux";
import { render, screen } from "@testing-library/react";
import Leaderboard from "../components/Leaderboard";
import configureStore from "redux-mock-store";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import kirby_brown from "../assets/kirby_brown.png";
import kirby_orange from "../assets/kirby_orange.png";
import kirby_green from "../assets/kirby_green.png";
import kirby_purple from "../assets/kirby_purple.png";

const mockStore = configureStore([]);

describe("Leaderboard Component", () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      users: {
        sarahedo: {
          id: "sarahedo",
          password: "password123",
          name: "Sarah Edo",
          avatarURL: kirby_brown,
          answers: {
            "8xf0y6ziyjabvozdd253nd": "optionOne",
            "6ni6ok3ym7mf1p33lnez": "optionOne",
            am8ehyc8byjqgar0jgpub9: "optionTwo",
            loxhs1bqm25b708cmbf3g: "optionTwo",
          },
          questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
        },
        tylermcginnis: {
          id: "tylermcginnis",
          password: "abc321",
          name: "Tyler McGinnis",
          avatarURL: kirby_green,
          answers: {
            vthrdm985a262al8qx3do: "optionOne",
            xj352vofupe1dqz9emx13r: "optionTwo",
          },
          questions: ["loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do"],
        },
        mtsamis: {
          id: "mtsamis",
          password: "xyz123",
          name: "Mike Tsamis",
          avatarURL: kirby_orange,
          answers: {
            xj352vofupe1dqz9emx13r: "optionOne",
            vthrdm985a262al8qx3do: "optionTwo",
            "6ni6ok3ym7mf1p33lnez": "optionOne",
          },
          questions: ["6ni6ok3ym7mf1p33lnez", "xj352vofupe1dqz9emx13r"],
        },
        zoshikanlu: {
          id: "zoshikanlu",
          password: "pass246",
          name: "Zenobia Oshikanlu",
          avatarURL: kirby_purple,
          answers: {
            xj352vofupe1dqz9emx13r: "optionOne",
          },
          questions: [],
        },
      },
    });
  });

  it("will display the correct name/qCount/aCount for each user", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Leaderboard />
        </MemoryRouter>
      </Provider>
    );

    // Sarah's values
    expect(screen.getByTestId("sarahedo-name").textContent).toEqual(
      "Sarah Edo"
    );
    expect(screen.getByTestId("sarahedo-questions").textContent).toEqual("2");
    expect(screen.getByTestId("sarahedo-answers").textContent).toEqual("4");

    // Tyler's Values
    expect(screen.getByTestId("tylermcginnis-name").textContent).toEqual(
      "Tyler McGinnis"
    );
    expect(screen.getByTestId("tylermcginnis-questions").textContent).toEqual(
      "2"
    );
    expect(screen.getByTestId("tylermcginnis-answers").textContent).toEqual(
      "2"
    );

    // Mike's Values
    expect(screen.getByTestId("mtsamis-name").textContent).toEqual(
      "Mike Tsamis"
    );
    expect(screen.getByTestId("mtsamis-questions").textContent).toEqual("2");
    expect(screen.getByTestId("mtsamis-answers").textContent).toEqual("3");

    // Zenobia's Values
    expect(screen.getByTestId("zoshikanlu-name").textContent).toEqual(
      "Zenobia Oshikanlu"
    );
    expect(screen.getByTestId("zoshikanlu-questions").textContent).toEqual("0");
    expect(screen.getByTestId("zoshikanlu-answers").textContent).toEqual("1");
  });

  it("will match the snapshot", () => {
    // eslint-disable-next-line testing-library/render-result-naming-convention
    var component = render(
      <Provider store={store}>
        <MemoryRouter>
          <Leaderboard />
        </MemoryRouter>
      </Provider>
    );

    expect(component).toMatchSnapshot();
  });
});
