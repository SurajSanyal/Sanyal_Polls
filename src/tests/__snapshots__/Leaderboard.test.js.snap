// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard Component will match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded mx-auto bg-slate-100 text-lg font-semibold py-2"
      >
        <table
          class="table-auto mx-auto w-full text-sm text-left text-blue-500"
        >
          <thead
            class="text-xs text-blue-500 uppercase bg-slate-100"
          >
            <tr>
              <th
                class="px-6 py-3"
              >
                User
              </th>
              <th
                class="px-6 py-3"
              >
                Questions Asked
              </th>
              <th
                class="px-6 py-3"
              >
                Questions Answered
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="bg-slate-100"
            >
              <td
                class="px-6 py-4 font-medium text-blue-500 whitespace-nowrap"
              >
                <div
                  class="flex items-center justify-self-auto"
                >
                  <img
                    alt="Profile"
                    class="h-10 pr-3"
                    data-testid="kirby_brown.png"
                    src="kirby_brown.png"
                  />
                  <span
                    data-testid="sarahedo-name"
                  >
                    Sarah Edo
                  </span>
                </div>
              </td>
              <td
                class="px-6 py-4"
                data-testid="sarahedo-questions"
              >
                2
              </td>
              <td
                class="px-6 py-4"
                data-testid="sarahedo-answers"
              >
                4
              </td>
            </tr>
            <tr
              class="bg-slate-100"
            >
              <td
                class="px-6 py-4 font-medium text-blue-500 whitespace-nowrap"
              >
                <div
                  class="flex items-center justify-self-auto"
                >
                  <img
                    alt="Profile"
                    class="h-10 pr-3"
                    data-testid="kirby_orange.png"
                    src="kirby_orange.png"
                  />
                  <span
                    data-testid="mtsamis-name"
                  >
                    Mike Tsamis
                  </span>
                </div>
              </td>
              <td
                class="px-6 py-4"
                data-testid="mtsamis-questions"
              >
                2
              </td>
              <td
                class="px-6 py-4"
                data-testid="mtsamis-answers"
              >
                3
              </td>
            </tr>
            <tr
              class="bg-slate-100"
            >
              <td
                class="px-6 py-4 font-medium text-blue-500 whitespace-nowrap"
              >
                <div
                  class="flex items-center justify-self-auto"
                >
                  <img
                    alt="Profile"
                    class="h-10 pr-3"
                    data-testid="kirby_green.png"
                    src="kirby_green.png"
                  />
                  <span
                    data-testid="tylermcginnis-name"
                  >
                    Tyler McGinnis
                  </span>
                </div>
              </td>
              <td
                class="px-6 py-4"
                data-testid="tylermcginnis-questions"
              >
                2
              </td>
              <td
                class="px-6 py-4"
                data-testid="tylermcginnis-answers"
              >
                2
              </td>
            </tr>
            <tr
              class="bg-slate-100"
            >
              <td
                class="px-6 py-4 font-medium text-blue-500 whitespace-nowrap"
              >
                <div
                  class="flex items-center justify-self-auto"
                >
                  <img
                    alt="Profile"
                    class="h-10 pr-3"
                    data-testid="kirby_purple.png"
                    src="kirby_purple.png"
                  />
                  <span
                    data-testid="zoshikanlu-name"
                  >
                    Zenobia Oshikanlu
                  </span>
                </div>
              </td>
              <td
                class="px-6 py-4"
                data-testid="zoshikanlu-questions"
              >
                0
              </td>
              <td
                class="px-6 py-4"
                data-testid="zoshikanlu-answers"
              >
                1
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded mx-auto bg-slate-100 text-lg font-semibold py-2"
    >
      <table
        class="table-auto mx-auto w-full text-sm text-left text-blue-500"
      >
        <thead
          class="text-xs text-blue-500 uppercase bg-slate-100"
        >
          <tr>
            <th
              class="px-6 py-3"
            >
              User
            </th>
            <th
              class="px-6 py-3"
            >
              Questions Asked
            </th>
            <th
              class="px-6 py-3"
            >
              Questions Answered
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="bg-slate-100"
          >
            <td
              class="px-6 py-4 font-medium text-blue-500 whitespace-nowrap"
            >
              <div
                class="flex items-center justify-self-auto"
              >
                <img
                  alt="Profile"
                  class="h-10 pr-3"
                  data-testid="kirby_brown.png"
                  src="kirby_brown.png"
                />
                <span
                  data-testid="sarahedo-name"
                >
                  Sarah Edo
                </span>
              </div>
            </td>
            <td
              class="px-6 py-4"
              data-testid="sarahedo-questions"
            >
              2
            </td>
            <td
              class="px-6 py-4"
              data-testid="sarahedo-answers"
            >
              4
            </td>
          </tr>
          <tr
            class="bg-slate-100"
          >
            <td
              class="px-6 py-4 font-medium text-blue-500 whitespace-nowrap"
            >
              <div
                class="flex items-center justify-self-auto"
              >
                <img
                  alt="Profile"
                  class="h-10 pr-3"
                  data-testid="kirby_orange.png"
                  src="kirby_orange.png"
                />
                <span
                  data-testid="mtsamis-name"
                >
                  Mike Tsamis
                </span>
              </div>
            </td>
            <td
              class="px-6 py-4"
              data-testid="mtsamis-questions"
            >
              2
            </td>
            <td
              class="px-6 py-4"
              data-testid="mtsamis-answers"
            >
              3
            </td>
          </tr>
          <tr
            class="bg-slate-100"
          >
            <td
              class="px-6 py-4 font-medium text-blue-500 whitespace-nowrap"
            >
              <div
                class="flex items-center justify-self-auto"
              >
                <img
                  alt="Profile"
                  class="h-10 pr-3"
                  data-testid="kirby_green.png"
                  src="kirby_green.png"
                />
                <span
                  data-testid="tylermcginnis-name"
                >
                  Tyler McGinnis
                </span>
              </div>
            </td>
            <td
              class="px-6 py-4"
              data-testid="tylermcginnis-questions"
            >
              2
            </td>
            <td
              class="px-6 py-4"
              data-testid="tylermcginnis-answers"
            >
              2
            </td>
          </tr>
          <tr
            class="bg-slate-100"
          >
            <td
              class="px-6 py-4 font-medium text-blue-500 whitespace-nowrap"
            >
              <div
                class="flex items-center justify-self-auto"
              >
                <img
                  alt="Profile"
                  class="h-10 pr-3"
                  data-testid="kirby_purple.png"
                  src="kirby_purple.png"
                />
                <span
                  data-testid="zoshikanlu-name"
                >
                  Zenobia Oshikanlu
                </span>
              </div>
            </td>
            <td
              class="px-6 py-4"
              data-testid="zoshikanlu-questions"
            >
              0
            </td>
            <td
              class="px-6 py-4"
              data-testid="zoshikanlu-answers"
            >
              1
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
